File: evaluations/matpower/tests/fnm_ingestion/test_g_fnm_4_fnm_acpf_convergence.m
%% Test G-FNM-4: ACPF convergence test with DCPF warm-start and progressive relaxation
%%
%% Dimension: fnm_ingestion
%% Network: LARGE (FNM Annual S01 -- main island, pre-cleaned .mat)
%% Pass condition: No hard pass/fail gate. All outcomes are diagnostic findings.
%%   Record relaxation_level_achieved: 0%, 10%, 20%, or infeasible.
%% Tool: MATPOWER 8.1
%% Input path: matpower (fallback .mat file)

fprintf('\n=== G-FNM-4: ACPF Convergence on Cleaned FNM ===\n\n');

%% Setup MATPOWER
mp_root = '/workspace/evaluations/matpower/matpower8.1';
addpath(fullfile(mp_root, 'lib'));
addpath(fullfile(mp_root, 'data'));
addpath(fullfile(mp_root, 'mips', 'lib'));
addpath(fullfile(mp_root, 'mp-opt-model', 'lib'));
addpath(fullfile(mp_root, 'mptest', 'lib'));
addpath('/workspace/evaluations/matpower/tests/fnm_ingestion');

define_constants;

%% Load cleaned FNM case
fprintf('Loading cleaned FNM case...\n');
mat_path = '/workspace/data/fnm/reference/cleaned/fnm_main_island.mat';
mpc = loadcase(mat_path);
fprintf('  Buses: %d\n', size(mpc.bus, 1));
fprintf('  Branches: %d\n', size(mpc.branch, 1));
fprintf('  Generators: %d\n', size(mpc.gen, 1));
fprintf('  baseMVA: %.1f\n', mpc.baseMVA);

%% Step 1: Solve DCPF for warm-start angles
fprintf('\n--- Step 1: DCPF for warm-start angles ---\n');
mpopt = mpoption('verbose', 0, 'out.all', 0);

tic_dcpf = tic;
dcpf_results = rundcpf(mpc, mpopt);
dcpf_time = toc(tic_dcpf);

fprintf('  DCPF success: %d\n', dcpf_results.success);
fprintf('  DCPF wall clock: %.3f seconds\n', dcpf_time);

if ~dcpf_results.success
    fprintf('\nFATAL: DCPF did not converge. Cannot proceed with ACPF.\n');
    return
end

dcpf_angles = dcpf_results.bus(:, VA);
dcpf_init_mean_deg = mean(abs(dcpf_angles));
dcpf_init_max_abs_deg = max(abs(dcpf_angles));

fprintf('  DCPF init mean angle: %.4f deg\n', dcpf_init_mean_deg);
fprintf('  DCPF init max |angle|: %.4f deg\n', dcpf_init_max_abs_deg);

%% Step 2: ACPF at 0% relaxation (nominal thermal limits)
fprintf('\n--- Step 2: ACPF at 0%% relaxation (nominal limits) ---\n');

mpc_ac = mpc;
mpc_ac.bus(:, VM) = 1.0;
mpc_ac.bus(:, VA) = dcpf_angles;

% Newton-Raphson with generous iteration limit
mpopt_ac = mpoption('verbose', 0, 'out.all', 0, ...
                    'pf.alg', 'NR', ...
                    'pf.nr.max_it', 100, ...
                    'pf.tol', 1e-8);

[~, rss_txt] = system('grep VmHWM /proc/self/status');
rss_before = sscanf(rss_txt, 'VmHWM: %f') / 1024;

tic_ac = tic;
acpf_results_0 = runpf(mpc_ac, mpopt_ac);
acpf_time_0 = toc(tic_ac);

[~, rss_txt] = system('grep VmHWM /proc/self/status');
rss_after_0 = sscanf(rss_txt, 'VmHWM: %f') / 1024;

fprintf('  ACPF (0%% relax) success: %d\n', acpf_results_0.success);
fprintf('  Wall clock: %.3f seconds\n', acpf_time_0);
fprintf('  Peak RSS: %.1f MB\n', rss_after_0);

relaxation_achieved = 'none_yet';

if acpf_results_0.success
    vm_vals = acpf_results_0.bus(:, VM);
    non_flat = sum(abs(vm_vals - 1.0) > 1e-6);
    non_flat_pct = non_flat / length(vm_vals) * 100;
    fprintf('  Non-flat VM buses: %d / %d (%.1f%%)\n', non_flat, length(vm_vals), non_flat_pct);
    fprintf('  VM range: [%.4f, %.4f] pu\n', min(vm_vals), max(vm_vals));
    fprintf('  VA range: [%.4f, %.4f] deg\n', min(acpf_results_0.bus(:, VA)), ...
            max(acpf_results_0.bus(:, VA)));

    total_gen_p = sum(acpf_results_0.gen(:, PG));
    total_load_p = sum(acpf_results_0.bus(:, PD));
    total_loss = total_gen_p - total_load_p;
    loss_pct = total_loss / total_load_p * 100;
    fprintf('  Total gen P: %.1f MW\n', total_gen_p);
    fprintf('  Total load P: %.1f MW\n', total_load_p);
    fprintf('  Total losses: %.1f MW (%.2f%%)\n', total_loss, loss_pct);

    relaxation_achieved = '0%';
    fprintf('\n  ACPF CONVERGED at 0%% relaxation.\n');
else
    fprintf('  ACPF did NOT converge at 0%% relaxation.\n');
end

%% Step 3: ACPF at 10% relaxation (if Step 2 failed)
if strcmp(relaxation_achieved, 'none_yet')
    fprintf('\n--- Step 3: ACPF at 10%% relaxation (RATE_A x 1.10) ---\n');

    mpc_ac_10 = mpc;
    mpc_ac_10.bus(:, VM) = 1.0;
    mpc_ac_10.bus(:, VA) = dcpf_angles;
    mpc_ac_10.branch(:, RATE_A) = mpc.branch(:, RATE_A) * 1.10;

    tic_ac_10 = tic;
    acpf_results_10 = runpf(mpc_ac_10, mpopt_ac);
    acpf_time_10 = toc(tic_ac_10);

    [~, rss_txt] = system('grep VmHWM /proc/self/status');
    rss_after_10 = sscanf(rss_txt, 'VmHWM: %f') / 1024;

    fprintf('  ACPF (10%% relax) success: %d\n', acpf_results_10.success);
    fprintf('  Wall clock: %.3f seconds\n', acpf_time_10);
    fprintf('  Peak RSS: %.1f MB\n', rss_after_10);

    if acpf_results_10.success
        vm_vals = acpf_results_10.bus(:, VM);
        non_flat = sum(abs(vm_vals - 1.0) > 1e-6);
        fprintf('  Non-flat VM buses: %d / %d (%.1f%%)\n', non_flat, length(vm_vals), ...
                non_flat / length(vm_vals) * 100);
        fprintf('  VM range: [%.4f, %.4f] pu\n', min(vm_vals), max(vm_vals));

        relaxation_achieved = '10%';
        fprintf('\n  ACPF CONVERGED at 10%% relaxation.\n');
    else
        fprintf('  ACPF did NOT converge at 10%% relaxation.\n');
    end
end

%% Step 4: ACPF at 20% relaxation (if Step 3 failed)
if strcmp(relaxation_achieved, 'none_yet')
    fprintf('\n--- Step 4: ACPF at 20%% relaxation (RATE_A x 1.20) ---\n');

    mpc_ac_20 = mpc;
    mpc_ac_20.bus(:, VM) = 1.0;
    mpc_ac_20.bus(:, VA) = dcpf_angles;
    mpc_ac_20.branch(:, RATE_A) = mpc.branch(:, RATE_A) * 1.20;

    tic_ac_20 = tic;
    acpf_results_20 = runpf(mpc_ac_20, mpopt_ac);
    acpf_time_20 = toc(tic_ac_20);

    [~, rss_txt] = system('grep VmHWM /proc/self/status');
    rss_after_20 = sscanf(rss_txt, 'VmHWM: %f') / 1024;

    fprintf('  ACPF (20%% relax) success: %d\n', acpf_results_20.success);
    fprintf('  Wall clock: %.3f seconds\n', acpf_time_20);
    fprintf('  Peak RSS: %.1f MB\n', rss_after_20);

    if acpf_results_20.success
        vm_vals = acpf_results_20.bus(:, VM);
        non_flat = sum(abs(vm_vals - 1.0) > 1e-6);
        fprintf('  Non-flat VM buses: %d / %d (%.1f%%)\n', non_flat, length(vm_vals), ...
                non_flat / length(vm_vals) * 100);
        fprintf('  VM range: [%.4f, %.4f] pu\n', min(vm_vals), max(vm_vals));

        relaxation_achieved = '20%';
        fprintf('\n  ACPF CONVERGED at 20%% relaxation.\n');
    else
        relaxation_achieved = 'infeasible';
        fprintf('\n  ACPF did NOT converge at 20%% relaxation.\n');
    end
end

%% Summary
fprintf('\n=== SUMMARY ===\n');
fprintf('DCPF warm-start: success\n');
fprintf('  dcpf_init_mean_deg: %.4f\n', dcpf_init_mean_deg);
fprintf('  dcpf_init_max_abs_deg: %.4f\n', dcpf_init_max_abs_deg);
fprintf('relaxation_level_achieved: %s\n', relaxation_achieved);
fprintf('Status: informational (no pass/fail gate)\n');

File: evaluations/powersimulations/tests/fnm_ingestion/test_g_fnm_4_fnm_acpf_convergence.jl
#=
Test G-FNM-4: ACPF Convergence (DCPF warm-start + progressive relaxation)

Dimension: fnm_ingestion
Network: LARGE (FNM main island via MATPOWER fallback, 28000 buses)
Pass condition: No hard gate. All outcomes are diagnostic.
  Record relaxation_level_achieved: 0%, 10%, 20%, or infeasible.
Tool: PowerSimulations.jl v0.30.2 (PowerSystems.jl v4.6.2, PowerFlows.jl v0.9.0)
=#

using Logging
using PowerSystems: PowerSystems
using PowerFlows: PowerFlows
using JSON: JSON

const PS = PowerSystems
const PF = PowerFlows

function peak_rss_mb()
    for line in eachline("/proc/self/status")
        if startswith(line, "VmHWM:")
            return parse(Float64, split(line)[2]) / 1024  # kB to MB
        end
    end
    return nothing
end

function run_test(;
    matpower_file::String="/workspace/data/fnm/reference/cleaned/fnm_main_island.m",
    acpf_timeout_minutes::Int=30,
)
    results = Dict(
        "status" => "informational",
        "wall_clock_seconds" => 0.0,
        "details" => Dict{String,Any}(),
        "errors" => String[],
        "workarounds" => String[],
    )

    # Suppress verbose logging
    logger = ConsoleLogger(stderr, Logging.Error)
    global_logger(logger)

    t0 = time()
    try
        # --- Step 1: Load network and solve DCPF for warm-start ---
        println("Step 1: Loading network and solving DCPF for warm-start...")
        t_load = time()
        sys = PS.System(matpower_file; runchecks=false)
        load_elapsed = time() - t_load
        println("  Network loaded in $(round(load_elapsed, digits=2))s")

        bus_count = length(collect(PS.get_components(PS.ACBus, sys)))
        println("  Bus count: $bus_count")

        println("  Solving DCPF...")
        t_dc = time()
        dc_result = PF.solve_powerflow(PF.DCPowerFlow(), sys)
        dc_elapsed = time() - t_dc
        println("  DCPF solved in $(round(dc_elapsed, digits=2))s")

        # Extract DCPF angles for warm-start diagnostics
        timestep_key = first(keys(dc_result))
        bus_df = dc_result[timestep_key]["bus_results"]

        # Bus angles are in radians
        angles_rad = bus_df.θ
        angles_deg = rad2deg.(angles_rad)
        nonzero_mask = abs.(angles_deg) .> 1e-10

        dcpf_init_mean_deg = mean(abs.(angles_deg[nonzero_mask]))
        dcpf_init_max_abs_deg = maximum(abs.(angles_deg))

        println("  DCPF warm-start angle stats:")
        println("    Mean |angle| (non-zero): $(round(dcpf_init_mean_deg, digits=4)) deg")
        println("    Max |angle|: $(round(dcpf_init_max_abs_deg, digits=4)) deg")
        println("    Non-zero angles: $(count(nonzero_mask)) / $(length(angles_deg))")

        results["details"]["dcpf_init_mean_deg"] = dcpf_init_mean_deg
        results["details"]["dcpf_init_max_abs_deg"] = dcpf_init_max_abs_deg
        results["details"]["dcpf_solve_seconds"] = dc_elapsed

        # Apply DCPF angles as warm-start to the System
        # PowerFlows.solve_powerflow! updates the system in-place after DCPF
        # We need to reload and set angles manually for AC
        println("\n  Setting DCPF angles on System buses for ACPF warm-start...")
        # Build angle map from DCPF results
        angle_map = Dict{Int,Float64}()
        for row in eachrow(bus_df)
            angle_map[row.bus_number] = row.θ  # radians
        end

        # Set bus angles from DCPF solution (VM stays at 1.0 pu for warm-start)
        for bus in PS.get_components(PS.ACBus, sys)
            bus_num = PS.get_number(bus)
            if haskey(angle_map, bus_num)
                PS.set_angle!(bus, angle_map[bus_num])
            end
            # Keep VM at 1.0 pu (default from MATPOWER load or generator setpoint)
        end

        # --- Step 2: ACPF at 0% relaxation ---
        println("\nStep 2: ACPF at 0% relaxation (no branch limit changes)...")
        relaxation_level_achieved = "infeasible"
        acpf_converged = false
        acpf_seconds = 0.0

        t_ac0 = time()
        try
            # Use ACPowerFlow with generous iteration limit
            converged = PF.solve_powerflow!(
                PF.ACPowerFlow(), sys; maxIterations=100, check_connectivity=false
            )
            acpf_seconds = time() - t_ac0

            if converged
                println("  ACPF converged at 0% relaxation in $(round(acpf_seconds, digits=2))s")
                relaxation_level_achieved = "0%"
                acpf_converged = true
            else
                println(
                    "  ACPF did NOT converge at 0% relaxation ($(round(acpf_seconds, digits=2))s)"
                )
            end
        catch e
            acpf_seconds = time() - t_ac0
            err_msg = string(typeof(e), ": ", sprint(showerror, e))
            println("  ACPF 0% error: $err_msg")
            push!(results["errors"], "ACPF 0%: $err_msg")
        end

        results["details"]["acpf_0pct_converged"] = acpf_converged
        results["details"]["acpf_0pct_seconds"] = acpf_seconds

        # --- Step 3: ACPF at 10% relaxation (if Step 2 failed) ---
        if !acpf_converged
            println("\nStep 3: ACPF at 10% relaxation...")
            println("  Reloading system and relaxing branch ratings by 10%...")

            # Reload system fresh
            sys = PS.System(matpower_file; runchecks=false)

            # Re-apply DCPF angles
            for bus in PS.get_components(PS.ACBus, sys)
                bus_num = PS.get_number(bus)
                if haskey(angle_map, bus_num)
                    PS.set_angle!(bus, angle_map[bus_num])
                end
            end

            # Relax branch ratings by 10%
            for branch in PS.get_components(PS.ACBranch, sys)
                rating = PS.get_rating(branch)
                if rating > 0
                    PS.set_rating!(branch, rating * 1.10)
                end
            end

            t_ac10 = time()
            try
                converged = PF.solve_powerflow!(
                    PF.ACPowerFlow(), sys; maxIterations=100, check_connectivity=false
                )
                ac10_seconds = time() - t_ac10

                if converged
                    println(
                        "  ACPF converged at 10% relaxation in $(round(ac10_seconds, digits=2))s"
                    )
                    relaxation_level_achieved = "10%"
                    acpf_converged = true
                else
                    println(
                        "  ACPF did NOT converge at 10% relaxation ($(round(ac10_seconds, digits=2))s)",
                    )
                end
                results["details"]["acpf_10pct_seconds"] = ac10_seconds
            catch e
                ac10_seconds = time() - t_ac10
                err_msg = string(typeof(e), ": ", sprint(showerror, e))
                println("  ACPF 10% error: $err_msg")
                push!(results["errors"], "ACPF 10%: $err_msg")
                results["details"]["acpf_10pct_seconds"] = ac10_seconds
            end

            results["details"]["acpf_10pct_converged"] = acpf_converged
        end

        # --- Step 4: ACPF at 20% relaxation (if Step 3 failed) ---
        if !acpf_converged
            println("\nStep 4: ACPF at 20% relaxation...")
            println("  Reloading system and relaxing branch ratings by 20%...")

            sys = PS.System(matpower_file; runchecks=false)

            for bus in PS.get_components(PS.ACBus, sys)
                bus_num = PS.get_number(bus)
                if haskey(angle_map, bus_num)
                    PS.set_angle!(bus, angle_map[bus_num])
                end
            end

            for branch in PS.get_components(PS.ACBranch, sys)
                rating = PS.get_rating(branch)
                if rating > 0
                    PS.set_rating!(branch, rating * 1.20)
                end
            end

            t_ac20 = time()
            try
                converged = PF.solve_powerflow!(
                    PF.ACPowerFlow(), sys; maxIterations=100, check_connectivity=false
                )
                ac20_seconds = time() - t_ac20

                if converged
                    println(
                        "  ACPF converged at 20% relaxation in $(round(ac20_seconds, digits=2))s"
                    )
                    relaxation_level_achieved = "20%"
                    acpf_converged = true
                else
                    println(
                        "  ACPF did NOT converge at 20% relaxation ($(round(ac20_seconds, digits=2))s)",
                    )
                end
                results["details"]["acpf_20pct_seconds"] = ac20_seconds
            catch e
                ac20_seconds = time() - t_ac20
                err_msg = string(typeof(e), ": ", sprint(showerror, e))
                println("  ACPF 20% error: $err_msg")
                push!(results["errors"], "ACPF 20%: $err_msg")
                results["details"]["acpf_20pct_seconds"] = ac20_seconds
            end

            results["details"]["acpf_20pct_converged"] = acpf_converged
        end

        results["details"]["relaxation_level_achieved"] = relaxation_level_achieved
        results["details"]["acpf_converged"] = acpf_converged
        results["details"]["peak_rss_mb"] = peak_rss_mb()

        println("\n=== Summary ===")
        println("  Relaxation level achieved: $relaxation_level_achieved")
        println("  ACPF converged: $acpf_converged")

    catch e
        push!(results["errors"], string(typeof(e), ": ", e))
        results["details"]["traceback"] = sprint(showerror, e, catch_backtrace())
        println("\n*** ERROR ***")
        println(results["details"]["traceback"])
    end

    results["wall_clock_seconds"] = time() - t0
    return results
end

# Helper: mean of a vector
function mean(x)
    return sum(x) / length(x)
end

function eachrow(df)
    return DataFrames.eachrow(df)
end

# Need DataFrames for eachrow
using DataFrames: DataFrames

if abspath(PROGRAM_FILE) == @__FILE__
    result = run_test()
    println("\n" * JSON.json(result, 2))
end
